

#Answer 2
##part(A)
To find the minimum value of $q_2(x)$, we have already derived the expression for $q_2(x)$:

$q_2(x) = 2q_1(x_1) + (2q_2 - 1)x_1 + \left(\frac{3}{2}q_2^2 - 2q_2\right)$

We can rewrite this expression in matrix form as:

$q_2(x) = \frac{1}{2}x^\top Ax - b_2^\top x$

Where:

$A = \begin{pmatrix} 4 & 1 \\ 1 & 3 \end{pmatrix}$

$b_2 = \begin{pmatrix} 1 \\ 2 \end{pmatrix}$

The minimum value of $q_2(x)$ occurs at the minimizer, which we can find by setting the gradient of $q_2(x)$ with respect to $x$ equal to zero:

$\nabla q_2(x) = 0$

$\Rightarrow Ax - b_2 = 0$

Solving this equation gives us the minimizer $x^*$:

$x^* = A^{-1}b_2$

We already know that $A$ is invertible, so we can compute $A^{-1}$ and find $x^*$.

Once we have $x^*$, we can substitute it into $q_2(x)$ to find the minimum value of $q_2(x)$.

Let's proceed with these calculations. First, let's find $A^{-1}$:

$A^{-1} = \frac{1}{\text{det}(A)} \begin{pmatrix} 3 & -1 \\ -1 & 4 \end{pmatrix}$

where $\text{det}(A) = 11$.

So,

$A^{-1} = \frac{1}{11} \begin{pmatrix} 3 & -1 \\ -1 & 4 \end{pmatrix} = \begin{pmatrix} \frac{3}{11} & -\frac{1}{11} \\ -\frac{1}{11} & \frac{4}{11} \end{pmatrix}$

Now,

$x^* = A^{-1}b_2 = \begin{pmatrix} \frac{3}{11} & -\frac{1}{11} \\ -\frac{1}{11} & \frac{4}{11} \end{pmatrix} \begin{pmatrix} 1 \\ 2 \end{pmatrix} = \begin{pmatrix} \frac{1}{11} \\ \frac{6}{11} \end{pmatrix}$

Now, substitute $x^*$ into $q_2(x)$ to find the minimum value:

$q_2(x^*) = 2 \begin{pmatrix} \frac{1}{11} \end{pmatrix}^\top A \begin{pmatrix} \frac{1}{11} \end{pmatrix} - \begin{pmatrix} 1 & 2 \end{pmatrix} \begin{pmatrix} \frac{1}{11} \\ \frac{6}{11} \end{pmatrix}$

$= 2 \left( \frac{1}{11} \right)^2 \begin{pmatrix} 4 & 1 \\ 1 & 3 \end{pmatrix} - \left( \frac{1}{11} + \frac{12}{11} \right)$

$= \frac{1}{121} \begin{pmatrix} 8 & 2 \\ 2 & 6 \end{pmatrix} - \frac{13}{11}$

$= \frac{1}{121} \begin{pmatrix} 8 & 2 \\ 2 & 6 \end{pmatrix} - \frac{13}{11}$

$= \frac{8}{121} + \frac{6}{121} - \frac{13}{11}$

$= \frac{57}{121}$

So, the minimum value of $q_2(x)$ is $\frac{57}{121}$.

Given:

$b_1 = \begin{pmatrix} 1 \\ 0 \end{pmatrix}$

The function $q_1(x)$ is:

$q_1(x) = \frac{1}{2} x^\top Wx - b_1^\top x$

Substituting $b_1$ into $q_1(x)$:

$q_1(x) = \frac{1}{2} x^\top Wx - \begin{pmatrix} 1 & 0 \end{pmatrix} \begin{pmatrix} x_1 \\ x_2 \end{pmatrix}$

$= \frac{1}{2} x^\top Wx - x_1$

Now, let's find the gradient of $q_1(x)$ with respect to $x$:

$\nabla q_1(x) = \frac{\partial}{\partial x} \left( \frac{1}{2} x^\top Wx - x_1 \right)$

$= \frac{1}{2} (W + W^\top) x - \begin{pmatrix} 1 \\ 0 \end{pmatrix}$

Setting the gradient to zero:

$\frac{1}{2} (W + W^\top) x - \begin{pmatrix} 1 \\ 0 \end{pmatrix} = 0$

$(W + W^\top) x = 2 \begin{pmatrix} 1 \\ 0 \end{pmatrix}$

Now, we can substitute the given value of $W$ into the equation and solve for $x^*$:

$\begin{pmatrix} 2t & 2\sqrt{t} \\ 2\sqrt{t} & 2+2t \end{pmatrix} x = 2 \begin{pmatrix} 1 \\ 0 \end{pmatrix}$

$\begin{pmatrix} t & \sqrt{t} \\ \sqrt{t} & 1+t \end{pmatrix} x = \begin{pmatrix} 1 \\ 0 \end{pmatrix}$

This system can be written as two equations:

$t x_1 + \sqrt{t} x_2 = 1$

$\sqrt{t} x_1 + (1+t) x_2 = 0$

Solving these equations will give us the minimizer $x^*$. Once we have $x^*$, we can substitute it into $q_1(x)$ to find the minimum value of $q_1(x)$.

First, let's multiply the first equation by $t$ and subtract the second equation from it to eliminate $x_1$:

\begin{align*}
(t\cdot 1 - 0) - \left(t \cdot \frac{2}{3} - (1+t)\right)x_2 &= t - t^2x_2 - tx_2 - (1+t)x_2 \\
&= tx_2 - (1+t)x_2 \\
&= t - tx_2 = t(1 - x_2) = -tx_2
\end{align*}

So, substituting $x_2 = -\frac{t}{t}$ into the first equation, we find:

$
\frac{2}{t}x_1 = 2
$

Therefore, the minimizer $x^*$ is:

$
x^* = \begin{pmatrix} \frac{2}{t} \\ -\frac{t}{t} \end{pmatrix} = \begin{pmatrix} \frac{2}{t} \\ -1 \end{pmatrix}
$


The minimum value of $q_1(x)$ for $t \neq 0$ is $\frac{1}{t^2} - \frac{t}{2} - \sqrt{t} + 1$.

*** Is the minimiser Unique***

Here we Can oobserve that minimiser is Unique for both of the Functions

*** Are They local or global Minima***

For $q_1(x)$ and $q_2(x)$:

$q_1(x)$: This is a quadratic function of $x$. Since the coefficient matrix $W$ is symmetric positive definite, $q_1(x)$ is convex. In convex optimization, any local minimum is also a global minimum. Therefore, the minimum obtained for $q_1(x)$ is a global minimum.

$q_2(x)$: Similar to $q_1(x)$, $q_2(x)$ is also a quadratic function of $x$. Since $A$ is symmetric positive definite, $q_2(x)$ is also convex. Hence, any local minimum for $q_2(x)$ is a global minimum.

In summary, the minima obtained for both $q_1(x)$ and $q_2(x)$ are global minima.




***Are q1 and q2 are convex ***

To determine if the functions $q_1(x)$ and $q_2(x)$ are convex, we need to analyze their Hessian matrices. A function is convex if its Hessian matrix is positive semidefinite for all $x$ in its domain.

For $q_1(x)$:

$ q_1(x) = \frac{1}{2} x^\top Wx - b_1^\top x $

The Hessian matrix $H$ of $q_1(x)$ is simply $W$ since there is no cross-term in the quadratic form. Given the structure of $W$, it's not immediately clear if $W$ is positive semidefinite. However, the given matrix $W$ is generally not symmetric, which is a requirement for a positive semidefinite matrix.

For $q_2(x)$:

$ q_2(x) = \frac{1}{2} x^\top Ax - b_2^\top x $

Here, the Hessian matrix $H$ is $A$. Since $A$ is symmetric and positive definite, it is also positive semidefinite.

In conclusion:
- $q_1(x)$ may not be convex due to the nonsymmetric nature of $W$.
- $q_2(x)$ is convex since $A$ is symmetric and positive definite.

Therefore, only $q_2(x)$ is convex.








#Part(b)

In [1]:
import numpy as np

# Define the function q1(x)
def q1(x):
    W = np.array([[t, np.sqrt(t)], [np.sqrt(t), 1 + t]])
    b1 = np.array([1, 0])
    return 0.5 * np.dot(np.dot(x.T, W), x) - np.dot(b1.T, x)

# Define the gradient of q1(x)
def grad_q1(x):
    W = np.array([[t, np.sqrt(t)], [np.sqrt(t), 1 + t]])
    return np.dot(W + W.T, x) - np.array([1, 0])

# Define the step size and momentum parameters
def alpha(t):
    return 2 / (3 + np.sqrt(9 - 4 * t**2))

def beta(t):
    mu0 = (3 + np.sqrt(9 - 4 * t**2)) / (3 - np.sqrt(9 - 4 * t**2))
    return np.sqrt(mu0 - 1) / np.sqrt(mu0 + 1)

# Initialize parameters
x0 = np.array([3, 5])
t = 0.001
tau = 1e-5

# Nesterov accelerated gradient descent
x = x0
x_prev = x0
k = 0
while np.linalg.norm(grad_q1(x)) > tau:
    grad_pert = grad_q1(x + beta(t) * (x - x_prev))
    x_next = x - alpha(t) * grad_pert + beta(t) * (x - x_prev)
    x_prev = x
    x = x_next
    k += 1

# Output the result
print("Optimal solution x*:", x)
print("Number of iterations:", k)


Optimal solution x*: [500499.95734358 -15811.38695327]
Number of iterations: 175243


In [2]:
import numpy as np

# Define the function q1(x)
def q1(x):
    W = np.array([[t, np.sqrt(t)], [np.sqrt(t), 1 + t]])
    b1 = np.array([1, 0])
    return 0.5 * np.dot(np.dot(x.T, W), x) - np.dot(b1.T, x)

# Define the gradient of q1(x)
def grad_q1(x):
    W = np.array([[t, np.sqrt(t)], [np.sqrt(t), 1 + t]])
    return np.dot(W + W.T, x) - np.array([1, 0])

# Define the step size
def alpha(t):
    return 2 / (3 + np.sqrt(9 - 4 * t**2))

# Initialize parameters
x0 = np.array([3, 5])
t = 0.001
tau = 1e-1
max_iter = 10000000

# Gradient descent
x = x0
k = 0
while np.linalg.norm(grad_q1(x)) > tau and k < max_iter:
    x = x - alpha(t) * grad_q1(x)
    k += 1

# Output the result
print("Optimal solution x*:", x)
print("Number of iterations:", k)


Optimal solution x*: [450425.05585283 -14229.45987838]
Number of iterations: 3460023


Let's compare the results obtained by both methods:

***Nesterov Accelerated Gradient Descent (NAGD):***

***Optimal solution*** $\mathbf{x^*}$: $[500499.95734358, -15811.38695327]$

***Number of iterations***: $175243$

***Gradient Descent:***

***Optimal solution*** $\mathbf{x^*}$: $[450425.05585283, -14229.45987838]$

***Number of iterations***: $3460023$

***Observations:***

***Optimal Solution:*** The optimal solution obtained by Nesterov Accelerated Gradient Descent (NAGD) is $[500499.95734358, -15811.38695327]$, while the one obtained by the gradient descent method is $[450425.05585283, -14229.45987838]$. The solution obtained by NAGD has higher values for both components of $\mathbf{x}$, indicating that it reaches a different local minimum compared to gradient descent.

***Number of Iterations:*** NAGD required $175243$ iterations to converge to its solution, while gradient descent required $3460023$ iterations. NAGD converged in significantly fewer iterations compared to gradient descent. This demonstrates the efficiency of Nesterov Accelerated Gradient Descent in converging to a solution.

***Convergence Behavior:*** The difference in the number of iterations and the obtained solutions suggests that NAGD might have found a more favorable optimization path or exploited the momentum parameter to escape local minima more effectively compared to gradient descent.

Overall, Nesterov Accelerated Gradient Descent seems to have provided a more efficient and effective optimization process, leading to a solution with higher values and in significantly fewer iterations compared to the basic gradient descent method.



***PArt(c)***

In [3]:
import numpy as np

# Define the function q2(x)
def q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return 0.5 * x.T.dot(A.dot(x)) - b2.T.dot(x)

# Define the gradient of q2(x)
def grad_q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return A.dot(x) - b2

# Nesterov's Accelerated Gradient Descent (NAGD) algorithm
def nesterov_accelerated_gradient_descent(x0, alpha, beta, tolerance):
    x = x0
    x_prev = x0
    t = 0
    while np.linalg.norm(grad_q2(x)) > tolerance:
        grad_pert = grad_q2(x + beta * (x - x_prev))
        x_next = x - alpha * grad_pert + beta * (x - x_prev)
        x_prev = x
        x = x_next
        t += 1
    return x, t

# Parameters
x0 = np.array([3, 5])
alpha = 2 / (7 + np.sqrt(5))
beta = np.sqrt((7 + np.sqrt(5)) / (7 - np.sqrt(5))) - 1 / np.sqrt((7 + np.sqrt(5)) / (7 - np.sqrt(5)))
tolerance = 1e-4

# Run NAGD algorithm
optimal_solution1, num_iterations1 = nesterov_accelerated_gradient_descent(x0, alpha, beta, tolerance)

# Print results
print("Optimal solution x*: ", optimal_solution1)
print("Number of iterations: ", num_iterations1)


Optimal solution x*:  [0.09089681 0.63638351]
Number of iterations:  18


In [11]:
# function for neta = 0.7
import numpy as np

# Define the function q2(x)
def q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return 0.5 * x.T.dot(A.dot(x)) - b2.T.dot(x)

# Define the gradient of q2(x)
def grad_q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return A.dot(x) - b2

# Nesterov's Accelerated Gradient Descent (NAGD) algorithm
def nesterov_accelerated_gradient_descent1(x0, alpha, tolerance):
    x = x0
    x_prev = x0
    t = 0
    while np.linalg.norm(grad_q2(x)) > tolerance:
        grad_pert = grad_q2(x + 0.7 * (x - x_prev))
        x_next = x - alpha * grad_pert + 0.7 * (x - x_prev)
        x_prev = x
        x = x_next
        t += 1
    return x, t

# Parameters
x0 = np.array([3, 5])
alpha = 2 / (7 + np.sqrt(5))

tolerance = 1e-4

# Run NAGD algorithm
optimal_solution2, num_iterations2 = nesterov_accelerated_gradient_descent1(x0, alpha, tolerance)

# Print results
print("Optimal solution x*: ", optimal_solution2)
print("Number of iterations: ", num_iterations2)

Optimal solution x*:  [0.09089743 0.63638251]
Number of iterations:  18


Here We can observe That for n=0.7 The results are same

In [4]:
import numpy as np

# Define the function q2(x)
def q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return 0.5 * x.T.dot(A.dot(x)) - b2.T.dot(x)

# Define the gradient of q2(x)
def grad_q2(x):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return A.dot(x) - b2

# Gradient Descent algorithm
def gradient_descent(x0, alpha, tolerance):
    x = x0
    t = 0
    while np.linalg.norm(grad_q2(x)) > tolerance:
        x -= alpha * grad_q2(x)
        t += 1
    return x, t

# Parameters
x0 = np.array([3.0, 5.0])
alpha1 = 2 / (7 + np.sqrt(5))
tolerance = 1e-2

# Run Gradient Descent algorithm
optimal_solution2, num_iterations2 = gradient_descent(x0, alpha1, tolerance)

# Print results
print("Optimal solution x*: ", optimal_solution2)
print("Number of iterations: ", num_iterations2)


Optimal solution x*:  [0.08923038 0.63907984]
Number of iterations:  9


Optimal Solution:

For Nesterov's accelerated gradient descent, the optimal solution is approximately
[0.09089681, 0.63638351].
For gradient descent, the optimal solution is approximately
[0.08923038, 0.63907984].
The solutions obtained by both algorithms are very close to each other, indicating that both methods effectively found a solution close to the minimum of the function q2(x).

Number of Iterations:

Nesterov's accelerated gradient descent required 18 iterations.
Gradient descent required only 9 iterations.
Gradient descent converged in fewer iterations compared to Nesterov's accelerated gradient descent. This suggests that gradient descent took larger steps towards the minimum of the function and hence converged faster.

Observations:

- Nesterov's accelerated gradient descent typically takes smaller steps compared to regular gradient descent due to the introduction of the momentum parameter. This can lead to slower convergence but may offer more stability, especially in scenarios with noisy or ill-conditioned objective functions.
- Gradient descent without momentum tends to take larger steps and might overshoot the minimum, but it converges faster in this particular scenario where the function is smooth and well-behaved.

Overall Behavior:

- Both algorithms effectively minimized the function q2(x) to a similar extent, yielding solutions that are very close.
- Nesterov's accelerated gradient descent showed a more cautious approach with smaller steps, resulting in a slightly slower convergence compared to gradient descent.
- Gradient descent, on the other hand, converged faster but might be more susceptible to oscillations or overshooting in scenarios with highly non-linear or noisy objective functions.


In [5]:
import numpy as np

def rosenbrock_function(x_vector):
    return 100*(x_vector[1]-x_vector[0]**2)**2 + (0.5-x_vector[0])**2

def gradient_rosenbrock(x_vector):
    return np.array([-400*x_vector[0]*(x_vector[1]-x_vector[0]**2)-2*(0.5-x_vector[0]), 200*(x_vector[1]-x_vector[0]**2) ])

def project_to_unit_ball(x_initial):
    if np.linalg.norm(x_initial) <= 1:
        return x_initial
    else:
        return x_initial / np.linalg.norm(x_initial)

def adaptive_gradient_descent(x_initial_guess, max_iterations, learning_rate):
    x_current = np.copy(x_initial_guess)
    x_sequence = []
    x_sequence.append(x_initial_guess)

    squared_gradient_norm_sum = np.linalg.norm(gradient_rosenbrock(x_current))**2

    for iteration in range(max_iterations):
        step_size = learning_rate / np.sqrt(squared_gradient_norm_sum)
        next_x = x_current - step_size * gradient_rosenbrock(x_current)

        x_current = project_to_unit_ball(next_x)
        squared_gradient_norm_sum += np.linalg.norm(gradient_rosenbrock(x_current))**2

        x_sequence.append(x_current)

    return x_current, rosenbrock_function(x_current), x_sequence

# Example usage:
initial_guess = np.array([1.0, 1.0])
max_iter = 1000
learning_rate = 0.1
result, minimum_value, sequence = adaptive_gradient_descent(initial_guess, max_iter, learning_rate)
print("Minimum point:", result)
print("Minimum value:", minimum_value)


Minimum point: [0.65954494 0.43576855]
Minimum value: 0.025513727642272492


In [6]:
x0 = np.array([0,0])
max_iter = [10**2, 500, 10**3, 5000, 10**4, 50000, 10**5, 500000, 10**6, 5000000]


for t in max_iter:
  R = 2
  minimizer, minimum, xks = adaptive_gradient_descent(x0, t, R)
  print('------------------------------------------------------')
  print('Max no of iterations = ',t)
  print('Minimizer = ',minimizer)
  print('Final value',minimum)

------------------------------------------------------
Max no of iterations =  100
Minimizer =  [-0.1191874   0.01564742]
Final value 0.383600910872348
------------------------------------------------------
Max no of iterations =  500
Minimizer =  [0.43711819 0.19075968]
Final value 0.003963895592601297
------------------------------------------------------
Max no of iterations =  1000
Minimizer =  [0.49185363 0.24187917]
Final value 6.652998985998079e-05
------------------------------------------------------
Max no of iterations =  5000
Minimizer =  [0.5  0.25]
Final value 2.207963373434313e-18
------------------------------------------------------
Max no of iterations =  10000
Minimizer =  [0.5  0.25]
Final value 1.262177448353619e-29
------------------------------------------------------
Max no of iterations =  50000
Minimizer =  [0.5  0.25]
Final value 1.262177448353619e-29
------------------------------------------------------
Max no of iterations =  100000
Minimizer =  [0.5  0.25

In [7]:
import numpy as np

def f1(x_vector):
    t = 5*x_vector[0] - 5
    m = 5*x_vector[1] - 5
    return np.sin(m)*np.exp((1-np.cos(t))**2) + np.cos(t)*np.exp((1-np.sin(m))**2) + (t-m)**2

def grad_f1(x_vector):
    t = 5*x_vector[0] - 5
    m = 5*x_vector[1] - 5
    return np.array([np.sin(m)*np.exp((1-np.cos(t))**2)*10*np.sin(t)*(1-np.cos(t)) - 5*np.exp((1-np.sin(m))**2)*np.sin(t) + 10*(t-m),
                     np.cos(m)*np.exp((1-np.cos(t))**2)*5 - 10*np.cos(t)*np.cos(m)*(1-np.sin(m))*np.exp((1-np.sin(m))**2) - 10*(t-m)])

def projection(x_initial):
    if np.linalg.norm(x_initial) <= 1:
        return x_initial
    else:
        return x_initial / np.linalg.norm(x_initial)

def adaptive_gradient_descent1(x_initial_guess, max_iterations, R_value):
    x_current = np.copy(x_initial_guess)
    x_sequence = []
    x_sequence.append(x_initial_guess)
    squared_grad_norm_sum = np.linalg.norm(grad_f1(x_current))**2

    for _ in range(max_iterations):
        alpha_value = R_value / np.sqrt(squared_grad_norm_sum)
        next_y = x_current - alpha_value * grad_f1(x_current)
        x_current = projection(next_y)
        squared_grad_norm_sum += np.linalg.norm(grad_f1(x_current))**2
        x_sequence.append(x_current)

    return x_current, f1(x_current), x_sequence

x_initial = np.array([0, 0])
max_iterations_list = [10**2, 500, 10**3, 5000, 10**4, 50000, 10**5, 500000, 10**6]
R_value = 2

for max_iter in max_iterations_list:
    minimizer, minimum_value, x_sequence = adaptive_gradient_descent1(x_initial, max_iter, R_value)
    print('------------------------------------------------------')
    print('Max number of iterations =', max_iter)
    print('Minimizer =', minimizer)
    print('Final value =', minimum_value)


------------------------------------------------------
Max number of iterations = 100
Minimizer = [-0.89293896 -0.45017776]
Final value = -67.7120096660121
------------------------------------------------------
Max number of iterations = 500
Minimizer = [-0.8381429  -0.54545071]
Final value = -97.89422078414144
------------------------------------------------------
Max number of iterations = 1000
Minimizer = [-0.8381429  -0.54545071]
Final value = -97.89422078414145
------------------------------------------------------
Max number of iterations = 5000
Minimizer = [-0.8381429  -0.54545071]
Final value = -97.89422078414145
------------------------------------------------------
Max number of iterations = 10000
Minimizer = [-0.8381429  -0.54545071]
Final value = -97.8942207841415
------------------------------------------------------
Max number of iterations = 50000
Minimizer = [-0.8381429  -0.54545071]
Final value = -97.89422078414145
------------------------------------------------------

In [8]:
import numpy as np

def q2(x_vector):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return 0.5 * x_vector @ A @ x_vector - b2 @ x_vector

def grad_q2(x_vector):
    A = np.array([[4, 1], [1, 3]])
    b2 = np.array([1, 2])
    return A @ x_vector - b2

def hess_q2():
    return np.array([[4, 1], [1, 3]])
def grad_descent(x0, alpha, tau):
    x = np.copy(x0)
    k = 0
    functional_values = []

    while np.linalg.norm(grad_q2(x)) > tau:
        x = x- alpha * grad_q2(x)
        k = k+1
        functional_values.append(q2(x))

    return k,x,q2(x),functional_values

def conjugate_gradient(x_initial_guess, tolerance):
    r = -1 * grad_q2(x_initial_guess)
    d = np.copy(r)
    x_current = np.copy(x_initial_guess)
    k = 0
    x_sequence = []

    while np.linalg.norm(r) > tolerance:
        alpha = r @ r / (d @ hess_q2() @ d)
        x_current += alpha * d
        r_next = r - alpha * (hess_q2() @ d)
        beta = -(r_next @ r_next) / (r @ r)
        d = r_next - beta * d
        r = np.copy(r_next)
        k += 1
        x_sequence.append(x_current)

    return x_current, q2(x_current), k

tolerance = 1e-8
initial_point = np.array([5., 3.])
minimizer_cg, final_value_cg, iterations_cg = conjugate_gradient(initial_point, tolerance)

# For comparison
alpha = 2 / (7 + np.sqrt(5))
iteration_gd, minimizer_gd, final_value_gd, function_value_gd = grad_descent(initial_point, alpha, tolerance)

print('Type = Conjugate Gradient')
print('Minimizer =', minimizer_cg)
print('Final value =', final_value_cg)
print('Number of iterations =', iterations_cg)
print('----------------------------------------------------------------------------------')
print('Type = Normal Gradient Descent')
print('Iteration =', iteration_gd)
print('Minimizer =', minimizer_gd)
print('Final value =', final_value_gd)
print('----------------------------------------------------------------------------------')


Type = Conjugate Gradient
Minimizer = [0.09090909 0.63636364]
Final value = -0.6818181818181819
Number of iterations = 2
----------------------------------------------------------------------------------
Type = Normal Gradient Descent
Iteration = 26
Minimizer = [0.09090909 0.63636363]
Final value = -0.6818181818181818
----------------------------------------------------------------------------------
